### PR TITLE
big refactor

### PR DESCRIPTION
**Pull request by BobTheBob9**

### CODE REVIEW

**Comment by BobTheBob9** [comment]
oh my fucking god does it commit the postbuild command

**Review by 0xMihir — CHANGES_REQUESTED** [review]
This post-build configuration is not necessary. You can just set the output folder to an environment variable, eg, `$(TitanfallInstallDir)`, and ask developers to set that up on their machine.  The docs on how to do this are here: https://docs.microsoft.com/en-us/visualstudio/msbuild/how-to-use-environment-variables-in-a-build?view=vs-2022

**Comment by BobTheBob9** [comment]
> This post-build configuration is not necessary. You can just set the output folder to an environment variable, eg, `$(TitanfallInstallDir)`, and ask developers to set that up on their machine. The docs on how to do this are here: https://docs.microsoft.com/en-us/visualstudio/msbuild/how-to-use-environment-variables-in-a-build?view=vs-2022

to be entirely honest i wasn't intending to have the postbuild step in this pr, but doesn't seem to be a way to gitignore a postbuild step or something like that
using the env var could be good tho

**Comment by F1F7Y** [comment]
Amongsus

File: NorthstarDLL/printcommand.h [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#pragma once
```
Comment (GeckoEidechse) [inline NorthstarDLL/printcommand.h:1]: Is there a reason the header is called `printcommand.h` (singular) while the C++ source file is called `printcommands` (plural)? ^^

File: loader_wsock32_proxy/loader_wsock32_proxy.vcxproj [hunk 2]
```diff
@@ -60,15 +60,15 @@
       <ConformanceMode>true</ConformanceMode>
       <PrecompiledHeader>Use</PrecompiledHeader>
       <PrecompiledHeaderFile>pch.h</PrecompiledHeaderFile>
-      <LanguageStandard>stdcpp17</LanguageStandard>
+      <LanguageStandard>stdcpp20</LanguageStandard>
       <AdditionalIncludeDirectories>..\NorthstarDedicatedTest\</AdditionalIncludeDirectories>
```
Comment (GeckoEidechse) [inline loader_wsock32_proxy/loader_wsock32_proxy.vcxproj:64]:
```suggestion
      <AdditionalIncludeDirectories>..\NorthstarDLL\</AdditionalIncludeDirectories>
```

This should probably be updated as well :P

File: NorthstarLauncher/main.cpp [hunk 3]
```diff
@@ -1,12 +1,17 @@
 #define WIN32_LEAN_AND_MEAN
-#include <Windows.h>
-#include <TlHelp32.h>
+#include <windows.h>
+#include <tlhelp32.h>
 #include <filesystem>
 #include <sstream>
 #include <fstream>
-#include <Shlwapi.h>
+ #include <shlwapi.h>
```
Comment (GeckoEidechse) [inline NorthstarLauncher/main.cpp:7]:
```suggestion
#include <shlwapi.h>
```

Accidental extra space

File: NorthstarLauncher/main.cpp [hunk 4]
```diff
@@ -147,7 +187,7 @@ void EnsureOriginStarted()
 		return;
 	}
 
-	printf("[*] Starting Origin...\n");
```
Comment (GeckoEidechse) [inline NorthstarLauncher/main.cpp:150]: Any particular reasons we're switching all the prints to `printf` in this file? ^^

File: loader_wsock32_proxy/loader_wsock32_proxy.vcxproj (hunk 2 above)
Comment (BobTheBob9) [reply loader_wsock32_proxy/loader_wsock32_proxy.vcxproj:64]: neither should be necessary, didn't notice we still had the include directories in the project file tho

File: NorthstarDLL/printcommand.h (hunk 1 above)
Comment (BobTheBob9) [reply NorthstarDLL/printcommand.h:1]: should probably be removed if there's nothing in it

File: NorthstarLauncher/main.cpp (hunk 4 above)
Comment (GeckoEidechse) [reply NorthstarLauncher/main.cpp:150]: Nevermind, got it wrong, we switched **from** `printf`s to C++ style prints.

Comment (GeckoEidechse) [reply NorthstarLauncher/main.cpp:150]: reason is, "cause we're in C++, not C"

File: NorthstarLauncher/NorthstarLauncher.vcxproj [hunk 5]
```diff
@@ -65,6 +65,9 @@
       <AdditionalDependencies>shlwapi.lib;kernel32.lib;user32.lib;gdi32.lib;winspool.lib;comdlg32.lib;advapi32.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;odbc32.lib;odbccp32.lib;%(AdditionalDependencies)</AdditionalDependencies>
       <StackReserveSize>8000000</StackReserveSize>
     </Link>
+    <PostBuildEvent>
+      <Command>copy /Y "$(TargetPath)" "$(SolutionDir)..\..\"</Command>
+    </PostBuildEvent>
```
Comment (GeckoEidechse) [inline NorthstarLauncher/NorthstarLauncher.vcxproj:70]: I think you wanna remove that still before merging, no?

File: NorthstarLauncher/NorthstarLauncher.vcxproj [hunk 6]
```diff
@@ -85,6 +88,9 @@
...
```
Comment (GeckoEidechse) [inline NorthstarLauncher/NorthstarLauncher.vcxproj:93]: I think you wanna remove that still before merging, no?

**Review by GeckoEidechse — CHANGES_REQUESTED** [review]

File: NorthstarLauncher/NorthstarLauncher.vcxproj (hunk 6 above)
Comment (BobTheBob9) [reply NorthstarLauncher/NorthstarLauncher.vcxproj:93]: these should be kept but I'll make them use an env var soontm

**Comment by BobTheBob9** [comment]
ok this shit ready i think

**Comment by pg9182** [comment]
Docker image for testing: `ghcr.io/pg9182/northstar-dedicated-ci:1-tf2.0.11.0-ns1.9.6-ci.59de24a5.4.3088619658-nslauncher.a15246e8.3086358793`

https://github.com/pg9182/northstar-dedicated-ci/actions/runs/3088619658

**Comment by x3Karma** [comment]
minor changes

**Review by emma-miler — APPROVED** [review]

File: NorthstarDLL/masterserver.cpp [hunk 7]
```diff
@@ -814,595 +692,385 @@ void MasterServerManager::AuthenticateWithServer(char* uid, char* playerToken, c
 	requestThread.detach();
 }
 
-void MasterServerManager::AddSelfToServerList(
-	int port, int authPort, char* name, char* description, char* map, char* playlist, int maxPlayers, char* password)
+void MasterServerManager::WritePlayerPersistentData(const char* playerId, const char* pdata, size_t pdataSize)
 {
-	if (!Cvar_ns_report_server_to_masterserver->GetBool())
```
Comment (RoyalBlue1) [inline NorthstarDLL/masterserver.cpp:820]: add this again

**Review by RoyalBlue1 — CHANGES_REQUESTED** [review]
Also stop retrying to connect to ms if it failed

File: NorthstarDLL/masterserver.cpp [hunk 8]
```diff
@@ -814,595 +692,385 @@ void MasterServerManager::AuthenticateWithServer(char* uid, char* playerToken, c
...
 
-				heartbeatThread.detach();
-			}
-			else
-			{
-				spdlog::error("Failed adding self to server list: error {}", curl_easy_strerror(result));
-				m_bSuccessfullyConnected = false;
-			}
+								if (serverAddedJson["error"].HasMember("enum") &&
+									!strcmp(serverAddedJson["error"]["enum"].GetString(), "DUPLICATE_SERVER"))
```
Comment (GeckoEidechse) [inline NorthstarDLL/masterserver.cpp:864]: As @RoyalBlue1 mentioned, it retries connection even when not `DUPLICATE_SERVER`. Not sure why this happens exactly though. ^^

Behaviour can be easily observed by just starting game, going into `mp_lobby` and watching console

**Review by GeckoEidechse — CHANGES_REQUESTED** [review]

**Review by GeckoEidechse — CHANGES_REQUESTED** [review]
Format check failures again :trollface:

File: NorthstarDLL/masterserver.cpp (hunk 8 above)
Comment (GeckoEidechse) [reply NorthstarDLL/masterserver.cpp:864]: Seems to be fixed somewhere between opening issue and b03070d5b4575581ff6183cada8083446caa351e

File: NorthstarDLL/audio.cpp [hunk 9]
```diff
@@ -324,7 +324,7 @@ void CustomAudioManager::ClearAudioOverrides()
 
 	// slightly (very) bad
 	// wait for all audio reads to complete so we don't kill preexisting audio buffers as we're writing to them
-	std::unique_lock lock(g_CustomAudioManager.m_loadingMutex);
+	std::unique_lock lock(m_loadingMutex);
```
Comment (GeckoEidechse) [inline NorthstarDLL/audio.cpp:327]: Wait that changes sth else that wasn't [the formatting issue](https://github.com/R2Northstar/NorthstarLauncher/actions/runs/3208788455/jobs/5244955961#step:4:20) yet now formatting check passes? wtf?

File: NorthstarDLL/masterserver.cpp (hunk 7 above)
Comment (GeckoEidechse) [reply NorthstarDLL/masterserver.cpp:820]: Seems to be working again, according to @RoyalBlue1 ^^

**Comment by GeckoEidechse** [comment]
TODOs:

- [ ] Regression check for token leak
- [ ] Regression check for other vulns
- [ ] ???

**Review by GeckoEidechse — CHANGES_REQUESTED** [review]
Fails connecting to any gameserver rn

**Comment by GeckoEidechse** [comment]
> Fails connecting to any gameserver rn

Fixed with 708bfedf71413bb2061a76c3ee4de5cbc8c56513

**Comment by pg9182** [comment]
[`ghcr.io/pg9182/northstar-dedicated-ci:1-tf2.0.11.0-ns1.9.7-ci.59de24a5.14.3211507256-nslauncher.708bfedf.3211209264-nsmods.1804b784`](https://github.com/pg9182/northstar-dedicated-ci/actions/runs/3211507256/attempts/1#summary-8782458325)

**Comment by pg9182** [comment]
`ghcr.io/pg9182/northstar-dedicated-ci:1-tf2.0.11.0-ns1.9.7-ci.59de24a5.16.3212510647-nslauncher.b281d360.3212449064-nsmods.2637aa4e`

Includea datatables from @BobTheBob9. Use `NS_EXTRA_ARGUMENTS=-nopakdedi`.

**Review by pg9182 — CHANGES_REQUESTED** [review]
Dedicated server fails to register with the master server. Tries to add itself multiple times (it's inconsistent); the first one works, but the other ones fail and breaks the auth process.

**Comment by Erlite** [comment]
> Dedicated server fails to register with the master server. Tries to add itself multiple times (it's inconsistent); the first one works, but the other ones fail and breaks the auth process.
> 
> ```
> ns-aitdm-1  | [02:09:14] [info] changing mod search path from z:\tmp\ns490603817\R2Northstar\mods\Northstar.Custom\mod to R2Northstar/mods\Northstar.CustomServers
> ns-aitdm-1  | [02:09:14] [info] SERVER CodeCallback CodeCallback_PostEntityInit called
> ns-aitdm-1  | [02:09:14] [info] Not executing custom callbacks for CodeCallback CodeCallback_PostEntityInit
> ns-aitdm-1  | [02:09:14] [info] [SERVER SCRIPT] Code Script: _init
> ns-aitdm-1  | [02:09:14] [info] loading took 15.629472999999999s
> ns-aitdm-1  | [02:09:15] [error] Failed reading masterserver response: got fastify error response
> ns-aitdm-1  | [02:09:15] [error] {"success":false,"error":{"enum":"DUPLICATE_SERVER","msg":"A server with this port already exists for your IP address"}}
> ns-aitdm-1  | [02:09:20] [error] Failed reading masterserver response: got fastify error response
> ns-aitdm-1  | [02:09:20] [error] {"success":false,"error":{"enum":"DUPLICATE_SERVER","msg":"A server with this port already exists for your IP address"}}
> ```
> 
> ```json
> {"time":"October 09 2022 02:09AM UTC","req":{"method":"POST","url":"/server/add_server?port=13314&authPort=13315&name=docker%209655a3f87b68%20testing%20server&description=go%20away&map=mp_colony02&playlist=aitdm&maxPlayers=12&password=","remotePort":40762},"msg":"incoming request"}
> {"time":"October 09 2022 02:09AM UTC","req":{"method":"POST","url":"/server/add_server?port=13314&authPort=13315&name=docker%209655a3f87b68%20testing%20server&description=go%20away&map=mp_colony02&playlist=aitdm&maxPlayers=12&password=","remotePort":10064},"msg":"incoming request"}
> {"time":"October 09 2022 02:09AM UTC","req":{"method":"POST","url":"/server/add_server?port=13314&authPort=13315&name=docker%209655a3f87b68%20testing%20server&description=go%20away&map=mp_colony02&playlist=aitdm&maxPlayers=12&password=","remotePort":64186},"msg":"incoming request"}
> {"time":"October 09 2022 02:09AM UTC","req":{"method":"POST","url":"/server/update_values?id=a4ccf7b6afe1d16288025f26fd2a0764&port=13314&authPort=13315&name=docker%209655a3f87b68%20testing%20server&description=go%20away&map=mp_colony02&playlist=aitdm&playerCount=0&maxPlayers=12&password=","remotePort":22456},"msg":"incoming request"}
> {"time":"October 09 2022 02:09AM UTC","req":{"method":"POST","url":"/server/update_values?id=a4ccf7b6afe1d16288025f26fd2a0764&port=13314&authPort=13315&name=docker%209655a3f87b68%20testing%20server&description=go%20away&map=mp_colony02&playlist=aitdm&playerCount=0&maxPlayers=12&password=","remotePort":40762},"msg":"incoming request"}
> {"time":"October 09 2022 02:09AM UTC","req":{"method":"POST","url":"/server/update_values?id=a4ccf7b6afe1d16288025f26fd2a0764&port=13314&authPort=13315&name=docker%209655a3f87b68%20testing%20server&description=go%20away&map=mp_colony02&playlist=aitdm&playerCount=0&maxPlayers=12&password=","remotePort":31646},"msg":"incoming request"}
> ```

Fix by the above PR.

**Comment by GeckoEidechse** [comment]
There seems to be an issue where other players cannot a refactor server and just "Authentication Failed" instead...

Error on client
```
[2022-10-17 00:57:32.764] [error] Failed reading masterserver response: got fastify error response
[2022-10-17 00:57:32.764] [error] {"success":false,"error":{"enum":"JSON_PARSE_ERROR","msg":"Error parsing json response"}}
```

**Comment by GeckoEidechse** [comment]
Also @emma-miler seems to have an issue with Northstar not launching properly on first run after cold start.

**Comment by GeckoEidechse** [comment]
Seems like update values has some big delay on refactor not a blocking issue but should get fixed before `v1.10` release.

**Review by GeckoEidechse — APPROVED** [review]
Apart from update values regression, I didn't notice any issues in testing with dedi docker server and around 4-8 players on Hetzner CPX21.

(Did not check code)

**Pull request merged**